### ISSUE TITLE
Points unevenly distributed with tries < 5

### ISSUE THREAD

**Issue by redblobgames** [kchapelier/fast-2d-poisson-disk-sampling#1]
I know a low number of `tries` may lead to uneven spacing but I'm trying to understand why it seems like the gaps are always in the lower right (x > 500, y > 500):

```js
function redraw() {
  let points = new FastPoissonDiskSampling({
    shape: [1000, 1000],
    radius: 5,
    tries: 4,
  }).fill();

  let ctx = document.querySelector('canvas').getContext('2d');
  ctx.clearRect(0, 0, 1000, 1000);
  ctx.fillStyle = "hsl(0, 50%, 50%)";
  for (let p of points) {
    ctx.fillRect(p[0], p[1], 2, 2);
  }
}

redraw();

```
[jsfiddle](https://jsfiddle.net/ntLypv9k/)

I will use tries=5 for my project to avoid this, but was curious what causes it. :-) Example outputs:

![tries-4-a](https://user-images.githubusercontent.com/8757369/167515893-8eb287d8-2010-42b3-9a57-39e179b60b18.png)
![tries-4b](https://user-images.githubusercontent.com/8757369/167515902-5b0dc324-f86a-47c6-8f4e-5aaca74135e3.png)

**Comment by kchapelier** [comment]
Good catch, I'll investigate this when I have some time.

**Comment by kchapelier** [comment]
Finally come around to investigating this. It seems to be related to a floating point precision issue when calculating the distance between the points to validate the candidates. But I don't really understand why only the points in this specific quadrant seems to be affected. The workaround is really simple, just changing the "epsilon" value ([here](https://github.com/kchapelier/fast-2d-poisson-disk-sampling/blob/master/src/fast-poisson-disk-sampling.js#L5)) from 2e-14 to 2e-12 fix the issue. I'll continue looking into it later today.

**Comment by kchapelier** [comment]
(broken english warning 😩)

Figured it out. I was almost sure this was related to the angles, because the test case with the 1000x1000 shape made it look like a quadrant thing (and because most of the algorithm implemented here is based on angles). This was off the mark. Testing the same thing in a 2000x2000 shape shows that the issue still happens when both coordinates are above 500 (most likely above 512).

This is just a matter of compounded float precision losses. The greater the absolute value of the float is, the lesser its precision. 

> 2 + Number.EPSILON * 7 // => 2.0000000000000018
> 4 + Number.EPSILON * 7 => 4.000000000000002
> 16 + Number.EPSILON * 7 => 16

Adding the cosinus of an angle to 512 "drastically" reduces its precision.

> 000 + Math.cos(1)  // => 000.5403023058681398
> 512 + Math.cos(1) // => 512.5403023058682

This happens everytime the coordinates of a point is calculated based on the coordinates of the previous point and its angle/orientation from it ([here](https://github.com/kchapelier/fast-2d-poisson-disk-sampling/blob/1.0.2/src/fast-poisson-disk-sampling.js#L187-L189)). The new point may end up with coordinates which are ever so slightly not far enough from the previous point due to this loss of precision.

The purpose of the existing "epsilon" variable was to mitigate the precision loss by slightly increasing the distance between the points.  But this "epsilon" value should be calculated based on the size of the space we work with, not naively hardcoded. Increasing the "epsilon" value like I mentionned in my previous message will just offset where the issue starts to manifest itself. Sure it may not happen after (512, 512) but after (2048, 2048) instead.

The fix will probably be something around those lines:

```js
var floatPrecisionMitigation = Math.max(1, Math.log2(Math.max(this.width, this.height)) - 8) | 0;
var epsilon = 2e-14 * Math.pow(2, floatPrecisionMitigation);
```

with an additional recommendation in the documentation to avoid using extremely large shape (10000x10000) and use a smaller radius instead

**Comment by redblobgames** [comment]
[I haven't looked closely at the algorithm, but speculating here] With floating point, it kind of feels like if I multiply the radius by 100 and also multiply the shape size by 100, I should get the same results but multiplied by 100, except for that epsilon affecting things. What if epsilon were relative to the shape size? Would we then get size independence?

**Comment by kchapelier** [comment]
> What if epsilon were relative to the shape size? Would we then get size independence?

Yes, that what I tried to prototype in the little blurb of code with "floatPrecisionMitigation". I'll still have to experiment to find the correct formula but the idea is to multiply epsilon by a power of two based on the log2 of the maximum coordinate of the shape. Say the shape is [1024, 256], we multiply epsilon by $2^{finetune(\lceil log2(1024) \rceil)}$.

**Comment by redblobgames** [comment]
Ahh, you're right, I wasn't reading the code closely. Does epsilon have to be a round number, or would something like `epsilon = Math.max(this.width, this.height) / 2.5e15` work as well?

**Comment by kchapelier** [comment]
I went with 

```js
var floatPrecisionMitigation = Math.max(1, Math.max(width, height) / 64 | 0);
var radiusEpsilon = 1e-14 * floatPrecisionMitigation;
```

(roughly similar to)

```
var floatPrecisionMitigation = Math.max(1, Math.log2(Math.max(width, height)) - 6);
var radiusEpsilon = 1e-14 * Math.pow(2, floatPrecisionMitigation);
```

Tested shapes' bounds from 512 to 512000000 without any issue. This change also allows the library to output resultsets with higher point density without making it any slower. Several false negatives due to FP precision issues are avoided.

While I was debugging the issue, I noticed some easy targets for performance optimization (leftovers of the arbitrary dimension handling from poisson-disk-sampling), so the library is about ~10% faster _with_ a higher point density now.

```
-- PREVIOUS (1.0.2) --

[1024x1024 radius 8 tries 4]: 11.007ms for ~11158 points
[1024x1024 radius 8 tries 15]: 28.878ms for ~13371 points
[1024x1024 radius 8 tries 30]: 44.346ms for ~14054 points
[1024x1024 radius 8 tries 60]: 77.559ms for ~14566 points

[1024x1024 radius 4 tries 4]: 44.730ms for ~43765 points
[1024x1024 radius 4 tries 15]: 114.411ms for ~53324 points
[1024x1024 radius 4 tries 30]: 186.775ms for ~56019 points
[1024x1024 radius 4 tries 60]: 324.102ms for ~58025 points

[1024x1024 radius 2 tries 4]: 195.887ms for ~164328 points
[1024x1024 radius 2 tries 15]: 486.990ms for ~212774 points
[1024x1024 radius 2 tries 30]: 783.791ms for ~223538 points
[1024x1024 radius 2 tries 60]: 1316.438ms for ~231541 points

-- FIX EPSILON --

[1024x1024 radius 8 tries 4]: 12.438ms for ~12210 points
[1024x1024 radius 8 tries 15]: 28.727ms for ~13688 points
[1024x1024 radius 8 tries 30]: 42.580ms for ~14361 points
[1024x1024 radius 8 tries 60]: 76.977ms for ~14786 points

[1024x1024 radius 4 tries 4]: 44.944ms for ~48684 points
[1024x1024 radius 4 tries 15]: 106.644ms for ~54461 points
[1024x1024 radius 4 tries 30]: 178.169ms for ~57189 points
[1024x1024 radius 4 tries 60]: 306.427ms for ~58803 points

[1024x1024 radius 2 tries 4]: 203.596ms for ~194453 points
[1024x1024 radius 2 tries 15]: 452.430ms for ~217548 points
[1024x1024 radius 2 tries 30]: 741.757ms for ~228146 points
[1024x1024 radius 2 tries 60]: 1263.042ms for ~234724 points

-- FIX EPSILON + OPTIMIZATIONS (1.0.3) --

[1024x1024 radius 8 tries 4]: 9.938ms for ~12214 points
[1024x1024 radius 8 tries 15]: 25.298ms for ~13703 points
[1024x1024 radius 8 tries 30]: 37.872ms for ~14354 points
[1024x1024 radius 8 tries 60]: 68.991ms for ~14765 points

[1024x1024 radius 4 tries 4]: 37.334ms for ~48697 points
[1024x1024 radius 4 tries 15]: 91.493ms for ~54517 points
[1024x1024 radius 4 tries 30]: 159.394ms for ~57174 points
[1024x1024 radius 4 tries 60]: 274.662ms for ~58824 points

[1024x1024 radius 2 tries 4]: 172.247ms for ~194463 points
[1024x1024 radius 2 tries 15]: 397.841ms for ~217559 points
[1024x1024 radius 2 tries 30]: 659.043ms for ~228213 points
[1024x1024 radius 2 tries 60]: 1134.083ms for ~234761 points
```
